cEP-0021.md

```diff
@@ -0,0 +1,178 @@
+  ...
+```
+
+`run_bear_from_section` method checks whether bear writer invoke the profiler
+interface or not i.e., keyword arguments passed to the Bear contain the
+`profile_bears` key or not.
+
+- If *not* then `run_bear_from_section()` method will directly call
+  the `run()` method of Bear so that coala can continue its normal
+  execution.
+- Else `run_bear_from_section()` method will call `profile()` method which
+  will check the whether the value of `profile_bears` key value is
+  `True` or `Dirname`.
+
+  - If `True` then `profile()` will create a profiler file for every
+    bear of every section with a name `{section_name}_{bear_name}.prof` and
+
+    dump it in a *current working directory.*
+
+    ```python
+    path =  os.getcwd()
+    if str(profile_bears)=='True':
+        filename = '{}_{}.prof'.format(
+            kwargs['section_name'], kwargs['bear_name'])
+        open(filename, 'wb')
+        kwargs.pop('section_name')
+        kwargs.pop('bear_name')
+        cProfile.runctx('run(*args, **kwargs)',
+                        globals(), locals(), filename)
+    ```
+
+  - If `Dirname` then `profile()` method will create a profiler file for
+    every bear of every section with a name `{section_name}_{bear_name}.prof`
+    and after that it check whether the specified directory exits or not.
+    If exists then it will directly dump the profiler files.
+    If not then it will create a directory with the literal name `Dirname` and
+    dump then profiler files on that.
+
+    ```python
+    subdir = str(profile_dump)
+    filepath = os.path.join(path, subdir)
+    if os.path.isdir(filepath):
+        filename = '{}_{}.prof'.format(
+            kwargs['section_name'], kwargs['bear_name'])
+        open(os.path.join(filepath,filename), 'wb')
+        kwargs.pop('section_name')
+        kwargs.pop('bear_name')
+        cProfile.runctx('self.run(*args, **kwargs)',
+                        globals(), locals(), os.path.join(filepath, filename))
+    else:
+        os.mkdir(os.path.join(path, subdir))
+        filename = '{}_{}.prof'.format(
+            kwargs['section_name'], kwargs['bear_name'])
+        open(os.path.join(filepath,filename), 'wb')
+        kwargs.pop('section_name')
+        kwargs.pop('bear_name')
+        cProfile.runctx('self.run(*args, **kwargs)',
+                        globals(), locals(), os.path.join(filepath, filename))
+    ```
+
+## Profiler Output
+
+The profiler produces binary output that can be displayed and formatted by
+third party tools, which will help bear writers in creating a better performant
+bear by analysing the profiler output.
+
+For e.g.
+
+With *CLI* arguments
+
+```console
+coala --b PEPE8Bear --files <file.py> --profile-bears
+```
+
+It will create a profile file with name `cli_PEP8Bear`.
+and after that Bear writers can use different modules like
+`pstats`, `snakeviz`, `cprofilev` etc. which reads profile results from a
+file and formats them in various ways i.e.
```

Comment:
Various ways..?
I asked for sample output and you gave that only for pstats. Why is it so hard to give sample output for snakeviz and cprofilev?
How do I install these programs, or read more about them.
Them seem to be critical for this projects success.

Reply:
@jayvdb ,snakeviz and cprofilev, both of then provide a local host link. Its not hard if you want I can save the htnl data, but don't that will property, as they work with live local host.

Reply:
>Various ways..?

I have already explained with taking `pstats` as an example.

Reply:
> How do I install these programs, or read more about them.

not done.

Reply:
was wating for your clearance :grinning:

Reply:
I am mentioning them because they are those modules which provide there known UI for visualization of profiler data, 
`pstats` is the core one only, both `SnakeViz` and `cprofilev` have is on `PyPI`

Reply:
`snakeviz` and `cprofilev` don't have `AppVeyor CI`, `snakeviz` support windows but `cprofilev` currenlty have one [opened](https://github.com/ymichael/cprofilev/issues/7) regarding windows.

Reply:
> I am mentioning them because they are those modules which provide there own UI for visualization of profiler data

How can the reader of this cEP know that unless you explain it in the cEP?

Compare snakeviz and cprofilev and provide brief summaries of what they provide for our bear writers and performance debuggers.

If they dont provide any useful features, you can say they exist, and what they do, and why they may not be considered useful for the coala bear use case.

Reply:
removing the `cprofilev` as there is already issue opened related to windows and no one has responded to that issue.

Reply:
Yup; we typically avoid unmaintained software, and software without good CI.